README.md
# Agile Jsmn

## 1、介绍

- `jsmn` 是一个超轻巧，携带方便，单文件，适用于单片机中存储空间有限的环境，简单的可以作为 ANSI-C 标准的 JSON 解析器。
- Agile Jsmn 在 jsmn 的基础上参照 cJSON 增加了一些 API 以实现简单的对象获取操作。

### 1.1、目录结构

| 名称 | 说明 |
| ---- | ---- |
| doc | 文档目录 |
| examples | 例程目录 |
| jsmn-1.0.0 | jsmn-1.0.0 源码目录 |
| util | Agile Jsmn 源码目录 |

### 1.2、许可证

Agile Jsmn 遵循 MIT 许可，详见 `LICENSE` 文件。

### 1.3、依赖

无依赖，使用标准 C 库开发，可直接在电脑上编译。

## 2、使用 Agile Jsmn

- 帮助文档请查看 [doc/doxygen/Agile_Jsmn.chm](./doc/doxygen/Agile_Jsmn.chm)

使用：

- jsmn_init 初始化 jsmn 解析器
- 使用 jsmn_parse 将 json 数据解析到 jsmntok_t 数组中
- JSMN_ItemInit 初始化根节点
- 调用相应 API 获取自己想要的 json 对象
- 值返回的是字符串

  如果值是整数或浮点数，需要自己使用 atoi 或 atof 转换

  如果需要判断值是不是 null，需要对字符串进行 strstr / strcmp 判断

### 2.1、示例

- 使用示例在 [examples](./examples) 下。

- 使用 `gcc` 可直接 `make` 编译，在电脑上运行。

### 2.2、Doxygen 文档生成

- 使用 `Doxywizard` 打开 [Doxyfile](./doc/doxygen/Doxyfile) 运行，生成的文件在 [doxygen/output](./doc/doxygen/output) 下。
- 需要更改 `Graphviz` 路径。
- `HTML` 生成未使用 `chm` 格式的，如果使能需要更改 `hhc.exe` 路径。

## 3、联系方式 & 感谢

- 维护：马龙伟
- 主页：<https://github.com/loogg/agile_jsmn>
- 邮箱：<2544047213@qq.com>
- jsmn 原作者主页：[https://github.com/zserge/jsmn](https://github.com/zserge/jsmn)


ChangeLog.md
# 更新记录

## Agile Jsmn 1.0.0 发布

### 新功能

2021-11-30：马龙伟

* 增加 Doxygen 注释，生成文档

### 修改

2021-11-30：马龙伟

* jsmn 源码包使用官方 1.0.0 的发布版
* jsmn 的配置 `JSMN_PARENT_LINKS` 在 `SConscript` 中使能

## Agile Jsmn 1.0.1 发布

### 修改

2021-12-29：马龙伟

* 修改 API 注释
* 完善文档
* 修改 example


inc/jsmn.h
#ifndef __JSMN_H_
#define __JSMN_H_

#include <stddef.h>

#ifdef __cplusplus
extern "C" {
#endif

#define JSMN_PARENT_LINKS

/**
 * JSON type identifier. Basic types are:
 *  o Object
 *  o Array
 *  o String
 *  o Other primitive: number, boolean (true/false) or null
 */
typedef enum
{
    JSMN_UNDEFINED = 0,
    JSMN_OBJECT = 1,
    JSMN_ARRAY = 2,
    JSMN_STRING = 3,
    JSMN_PRIMITIVE = 4
} jsmntype_t;

enum jsmnerr
{
    /* Not enough tokens were provided */
    JSMN_ERROR_NOMEM = -1,
    /* Invalid character inside JSON string */
    JSMN_ERROR_INVAL = -2,
    /* The string is not a full JSON packet, more bytes expected */
    JSMN_ERROR_PART = -3
};

/**
 * JSON token description.
 * type     type (object, array, string etc.)
 * start    start position in JSON data string
 * end      end position in JSON data string
 */
typedef struct
{
    jsmntype_t type;
    int start;
    int end;
    int size;
#ifdef JSMN_PARENT_LINKS
    int parent;
#endif
} jsmntok_t;

/**
 * JSON parser. Contains an array of token blocks available. Also stores
 * the string being parsed now and current position in that string
 */
typedef struct
{
    unsigned int pos; /* offset in the JSON string */
    unsigned int toknext; /* next token to allocate */
    int toksuper; /* superior token node, e.g parent object or array */
} jsmn_parser;

/**
 * Create JSON parser over an array of tokens
 */
void jsmn_init(jsmn_parser *parser);

/**
 * Run JSON parser. It parses a JSON data string into and array of tokens, each describing
 * a single JSON object.
 */
int jsmn_parse(jsmn_parser *parser, const char *js, size_t len,
               jsmntok_t *tokens, unsigned int num_tokens);

#ifdef __cplusplus
}
#endif

#endif /* __JSMN_H_ */


src/jsmn.c
#include "jsmn.h"

/**
 * Allocates a fresh unused token from the token pull.
 */
static jsmntok_t *jsmn_alloc_token(jsmn_parser *parser,
                                   jsmntok_t *tokens, size_t num_tokens)
{
    jsmntok_t *tok;
    if (parser->toknext >= num_tokens)
    {
        return NULL;
    }
    tok = &tokens[parser->toknext++];
    tok->start = tok->end = -1;
    tok->size = 0;
#ifdef JSMN_PARENT_LINKS
    tok->parent = -1;
#endif
    return tok;
}

/**
 * Fills token type and boundaries.
 */
static void jsmn_fill_token(jsmntok_t *token, jsmntype_t type,
                            int start, int end)
{
    token->type = type;
    token->start = start;
    token->end = end;
    token->size = 0;
}

/**
 * Fills next available token with JSON primitive.
 */
static int jsmn_parse_primitive(jsmn_parser *parser, const char *js,
                                size_t len, jsmntok_t *tokens, size_t num_tokens)
{
    jsmntok_t *token;
    int start;

    start = parser->pos;

    for (; parser->pos < len && js[parser->pos] != '\0'; parser->pos++)
    {
        switch (js[parser->pos])
        {
#ifndef JSMN_STRICT
        /* In strict mode primitive must be followed by "," or "}" or "]" */
        case ':':
#endif
        case '\t' :
        case '\r' :
        case '\n' :
        case ' ' :
        case ','  :
        case ']'  :
        case '}' :
            goto found;
        }
        if (js[parser->pos] < 32 || js[parser->pos] >= 127)
        {
            parser->pos = start;
            return JSMN_ERROR_INVAL;
        }
    }
#ifdef JSMN_STRICT
    /* In strict mode primitive must be followed by a comma/object/array */
    parser->pos = start;
    return JSMN_ERROR_PART;
#endif

found:
    if (tokens == NULL)
    {
        parser->pos--;
        return 0;
    }
    token = jsmn_alloc_token(parser, tokens, num_tokens);
    if (token == NULL)
    {
        parser->pos = start;
        return JSMN_ERROR_NOMEM;
    }
    jsmn_fill_token(token, JSMN_PRIMITIVE, start, parser->pos);
#ifdef JSMN_PARENT_LINKS
    token->parent = parser->toksuper;
#endif
    parser->pos--;
    return 0;
}

/**
 * Fills next token with JSON string.
 */
static int jsmn_parse_string(jsmn_parser *parser, const char *js,
                             size_t len, jsmntok_t *tokens, size_t num_tokens)
{
    jsmntok_t *token;

    int start = parser->pos;

    parser->pos++;

    /* Skip starting quote */
    for (; parser->pos < len && js[parser->pos] != '\0'; parser->pos++)
    {
        char c = js[parser->pos];

        /* Quote: end of string */
        if (c == '\"')
        {
            if (tokens == NULL)
            {
                return 0;
            }
            token = jsmn_alloc_token(parser, tokens, num_tokens);
            if (token == NULL)
            {
                parser->pos = start;
                return JSMN_ERROR_NOMEM;
            }
            jsmn_fill_token(token, JSMN_STRING, start + 1, parser->pos);
#ifdef JSMN_PARENT_LINKS
            token->parent = parser->toksuper;
#endif
            return 0;
        }

        /* Backslash: Quoted symbol expected */
        if (c == '\\' && parser->pos + 1 < len)
        {
            int i;
            parser->pos++;
            switch (js[parser->pos])
            {
            /* Allowed escaped symbols */
            case '\"':
            case '/' :
            case '\\' :
            case 'b' :
            case 'f' :
            case 'r' :
            case 'n'  :
            case 't' :
                break;
            /* Allows escaped symbol \uXXXX */
            case 'u':
                parser->pos++;
                for (i = 0; i < 4 && parser->pos < len && js[parser->pos] != '\0'; i++)
                {
                    /* If it isn't a hex character we have an error */
                    if (!((js[parser->pos] >= 48 && js[parser->pos] <= 57) || /* 0-9 */
                            (js[parser->pos] >= 65 && js[parser->pos] <= 70) || /* A-F */
                            (js[parser->pos] >= 97 && js[parser->pos] <= 102)))   /* a-f */
                    {
                        parser->pos = start;
                        return JSMN_ERROR_INVAL;
                    }
                    parser->pos++;
                }
                parser->pos--;
                break;
            /* Unexpected symbol */
            default:
                parser->pos = start;
                return JSMN_ERROR_INVAL;
            }
        }
    }
    parser->pos = start;
    return JSMN_ERROR_PART;
}

/**
 * Parse JSON string and fill tokens.
 */
int jsmn_parse(jsmn_parser *parser, const char *js, size_t len,
               jsmntok_t *tokens, unsigned int num_tokens)
{
    int r;
    int i;
    jsmntok_t *token;
    int count = parser->toknext;

    for (; parser->pos < len && js[parser->pos] != '\0'; parser->pos++)
    {
        char c;
        jsmntype_t type;

        c = js[parser->pos];
        switch (c)
        {
        case '{':
        case '[':
            count++;
            if (tokens == NULL)
            {
                break;
            }
            token = jsmn_alloc_token(parser, tokens, num_tokens);
            if (token == NULL)
                return JSMN_ERROR_NOMEM;
            if (parser->toksuper != -1)
            {
                tokens[parser->toksuper].size++;
#ifdef JSMN_PARENT_LINKS
                token->parent = parser->toksuper;
#endif
            }
            token->type = (c == '{' ? JSMN_OBJECT : JSMN_ARRAY);
            token->start = parser->pos;
            parser->toksuper = parser->toknext - 1;
            break;
        case '}':
        case ']':
            if (tokens == NULL)
                break;
            type = (c == '}' ? JSMN_OBJECT : JSMN_ARRAY);
#ifdef JSMN_PARENT_LINKS
            if (parser->toknext < 1)
            {
                return JSMN_ERROR_INVAL;
            }
            token = &tokens[parser->toknext - 1];
            for (;;)
            {
                if (token->start != -1 && token->end == -1)
                {
                    if (token->type != type)
                    {
                        return JSMN_ERROR_INVAL;
                    }
                    token->end = parser->pos + 1;
                    parser->toksuper = token->parent;
                    break;
                }
                if (token->parent == -1)
                {
                    if (token->type != type || parser->toksuper == -1)
                    {
                        return JSMN_ERROR_INVAL;
                    }
                    break;
                }
                token = &tokens[token->parent];
            }
#else
            for (i = parser->toknext - 1; i >= 0; i--)
            {
                token = &tokens[i];
                if (token->start != -1 && token->end == -1)
                {
                    if (token->type != type)
                    {
                        return JSMN_ERROR_INVAL;
                    }
                    parser->toksuper = -1;
                    token->end = parser->pos + 1;
                    break;
                }
            }
            /* Error if unmatched closing bracket */
            if (i == -1) return JSMN_ERROR_INVAL;
            for (; i >= 0; i--)
            {
                token = &tokens[i];
                if (token->start != -1 && token->end == -1)
                {
                    parser->toksuper = i;
                    break;
                }
            }
#endif
            break;
        case '\"':
            r = jsmn_parse_string(parser, js, len, tokens, num_tokens);
            if (r < 0) return r;
            count++;
            if (parser->toksuper != -1 && tokens != NULL)
                tokens[parser->toksuper].size++;
            break;
        case '\t' :
        case '\r' :
        case '\n' :
        case ' ':
            break;
        case ':':
            parser->toksuper = parser->toknext - 1;
            break;
        case ',':
            if (tokens != NULL && parser->toksuper != -1 &&
                    tokens[parser->toksuper].type != JSMN_ARRAY &&
                    tokens[parser->toksuper].type != JSMN_OBJECT)
            {
#ifdef JSMN_PARENT_LINKS
                parser->toksuper = tokens[parser->toksuper].parent;
#else
                for (i = parser->toknext - 1; i >= 0; i--)
                {
                    if (tokens[i].type == JSMN_ARRAY || tokens[i].type == JSMN_OBJECT)
                    {
                        if (tokens[i].start != -1 && tokens[i].end == -1)
                        {
                            parser->toksuper = i;
                            break;
                        }
                    }
                }
#endif
            }
            break;
#ifdef JSMN_STRICT
        /* In strict mode primitives are: numbers and booleans */
        case '-':
        case '0':
        case '1' :
        case '2':
        case '3' :
        case '4':
        case '5':
        case '6':
        case '7' :
        case '8':
        case '9':
        case 't':
        case 'f':
        case 'n' :
            /* And they must not be keys of the object */
            if (tokens != NULL && parser->toksuper != -1)
            {
                jsmntok_t *t = &tokens[parser->toksuper];
                if (t->type == JSMN_OBJECT ||
                        (t->type == JSMN_STRING && t->size != 0))
                {
                    return JSMN_ERROR_INVAL;
                }
            }
#else
        /* In non-strict mode every unquoted value is a primitive */
        default:
#endif
            r = jsmn_parse_primitive(parser, js, len, tokens, num_tokens);
            if (r < 0) return r;
            count++;
            if (parser->toksuper != -1 && tokens != NULL)
                tokens[parser->toksuper].size++;
            break;

#ifdef JSMN_STRICT
        /* Unexpected char in strict mode */
        default:
            return JSMN_ERROR_INVAL;
#endif
        }
    }

    if (tokens != NULL)
    {
        for (i = parser->toknext - 1; i >= 0; i--)
        {
            /* Unmatched opened object or array */
            if (tokens[i].start != -1 && tokens[i].end == -1)
            {
                return JSMN_ERROR_PART;
            }
        }
    }

    return count;
}

/**
 * Creates a new parser based over a given buffer with an array of tokens
 * available.
 */
void jsmn_init(jsmn_parser *parser)
{
    parser->pos = 0;
    parser->toknext = 0;
    parser->toksuper = -1;
}


util/jsmn_util.h
/**
 * @file    jsmn_util.h
 * @brief   Agile Jsmn 软件包头文件
 * @author  马龙伟 (2544047213@qq.com)
 * @version 1.0.1
 * @date    2021-12-29
 *
 * @attention
 *
 * <h2><center>&copy; Copyright (c) 2021 Ma Longwei.
 * All rights reserved.</center></h2>
 *
 */

#ifndef __JSMN_UTIL_H
#define __JSMN_UTIL_H

#ifdef __cplusplus
extern "C" {
#endif

#include "jsmn.h"

/** @defgroup AGILE_JSMN_Exported_Types Agile Jsmn Exported Types
 * @{
 */

/**
 * @brief   Agile Jsmn 结构体
 */
typedef struct
{
    jsmntok_t *t; /**< 对应 jsmn token 对象 */
    int index;    /**< 对应 jsmn token 对象在 token 数组中的索引 */
    int left_num; /**< token 数组中剩余 token 数目 */
} jsmn_item_t;

/**
 * @}
 */

/** @addtogroup AGILE_JSMN_Exported_Functions
 * @{
 */
void JSMN_ItemInit(jsmn_item_t *item, jsmntok_t *t, int index, int tokens_len);
int JSMN_GetObjectItem(const char *js, jsmn_item_t *object, const char *const string, jsmn_item_t *item);
char *JSMN_GetString(char *js, jsmn_item_t *item);
int JSMN_GetStringBuffered(const char *js, jsmn_item_t *item, char *buf, int bufsz);
char *JSMN_GetValueString(char *js, jsmn_item_t *item);
int JSMN_GetValueStringBuffered(const char *js, jsmn_item_t *item, char *buf, int bufsz);
int JSMN_GetArraySize(jsmn_item_t *array);
int JSMN_GetArrayItem(jsmn_item_t *array, int index, jsmn_item_t *item);
/**
 * @}
 */

#ifdef __cplusplus
}
#endif

#endif /* __JSMN_UTIL_H */


examples/example_jsmn.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "jsmn.h"
#include "jsmn_util.h"

#define DBG_ENABLE
#define DBG_COLOR
#define DBG_LEVEL        DBG_LOG
#define DBG_SECTION_NAME "example_jsmn"
#include "dbg_log.h"

#define EXAMPLE_FILE_PATH "example.json"

int main(void)
{
    system("cls");
    FILE *fp = fopen(EXAMPLE_FILE_PATH, "rb");
    if (!fp) {
        LOG_E("open file:%s error", EXAMPLE_FILE_PATH);
        return -1;
    }
    fseek(fp, 0, SEEK_END);
    int nFileLen = ftell(fp);
    fseek(fp, 0, SEEK_SET);
    char *js = (char *)malloc(nFileLen + 1);
    fread(js, nFileLen, 1, fp);
    fclose(fp);
    js[nFileLen] = 0;

    do {
        jsmn_parser parser;
        jsmn_init(&parser);
        jsmntok_t tokens[100] = {0};
        int tokens_len = jsmn_parse(&parser, js, strlen(js), tokens, sizeof(tokens) / sizeof(tokens[0]));
        if (tokens_len <= 0)
            break;
        jsmn_item_t root;
        JSMN_ItemInit(&root, tokens, 0, tokens_len);
        int root_array_size = JSMN_GetArraySize(&root);
        if (root_array_size < 0)
            break;
        for (int i = 0; i < root_array_size; i++) {
            jsmn_item_t root_ob_item;
            if (JSMN_GetArrayItem(&root, i, &root_ob_item) != 0)
                continue;

            jsmn_item_t root_ob_a_item;
            jsmn_item_t root_ob_b_item;
            if (JSMN_GetObjectItem(js, &root_ob_item, "a", &root_ob_a_item) != 0)
                continue;
            if (JSMN_GetObjectItem(js, &root_ob_item, "b", &root_ob_b_item) != 0)
                continue;

            char *item_string = JSMN_GetValueString(js, &root_ob_a_item);
            if (item_string == NULL)
                continue;

            LOG_I("get root array index:%d--->\"%s\":%s", i, JSMN_GetString(js, &root_ob_a_item), item_string);

            int b_array_size = JSMN_GetArraySize(&root_ob_b_item);
            if (b_array_size < 0)
                continue;

            for (int j = 0; j < b_array_size; j++) {
                jsmn_item_t b_ob_item;
                if (JSMN_GetArrayItem(&root_ob_b_item, j, &b_ob_item) != 0)
                    continue;

                jsmn_item_t b_ob_a_item;
                if (JSMN_GetObjectItem(js, &b_ob_item, "a", &b_ob_a_item) != 0)
                    continue;

                jsmn_item_t b_ob_a_value_item;
                jsmn_item_t b_ob_a_name_item;
                if (JSMN_GetObjectItem(js, &b_ob_a_item, "value", &b_ob_a_value_item) != 0)
                    continue;
                if (JSMN_GetObjectItem(js, &b_ob_a_item, "name", &b_ob_a_name_item) != 0)
                    continue;

                LOG_I("get root array index:%d--->\"%s\" array index:%d--->\"%s\":%s, \"%s\":%s", i, JSMN_GetString(js, &root_ob_b_item),
                      j, JSMN_GetString(js, &b_ob_a_value_item), JSMN_GetValueString(js, &b_ob_a_value_item), JSMN_GetString(js, &b_ob_a_name_item),
                      JSMN_GetValueString(js, &b_ob_a_name_item));
            }
        }
    } while (0);

    free(js);
    getchar();
    return 0;
}


util/jsmn_util.c
/**
 * @file    jsmn_util.c
 * @brief   Agile Jsmn 软件包源文件
 * @author  马龙伟 (2544047213@qq.com)
 * @version 1.0.1
 * @date    2021-12-29
 *
 @verbatim
    使用：
    - jsmn_init 初始化 jsmn 解析器
    - 使用 jsmn_parse 将 json 数据解析到 jsmntok_t 数组中
    - JSMN_ItemInit 初始化根节点
    - 调用相应 API 获取自己想要的 json 对象
    - 值返回的是字符串
      如果值是整数或浮点数，需要自己使用 atoi 或 atof 转换
      如果需要判断值是不是 null，需要对字符串进行 strstr / strcmp 判断

 @endverbatim
 *
 * @attention
 *
 * <h2><center>&copy; Copyright (c) 2021 Ma Longwei.
 * All rights reserved.</center></h2>
 *
 */

#include "jsmn_util.h"
#include <string.h>

/** @defgroup AGILE_JSMN_Private_Functions Agile Jsmn Private Functions
 * @{
 */

/**
 * @brief   string 类型 token 字符串比较
 * @param   js json 字符串
 * @param   t token 对象
 * @param   s 要比较的字符串
 * @return  1:相同; 0:不同
 */
static int json_token_streq(const char *js, jsmntok_t *t, const char *const s)
{
    return (strncmp(js + t->start, s, t->end - t->start) == 0 && strlen(s) == (size_t)(t->end - t->start));
}

/**
 * @}
 */

/** @defgroup AGILE_JSMN_Exported_Functions Agile Jsmn Exported Functions
 * @{
 */

/**
 * @brief   初始化 Agile Jsmn 对象
 * @param   item Agile Jsmn 对象指针
 * @param   t 对应的 token
 * @param   index 对应的 token 索引
 * @param   tokens_len 数组剩余 token 数目
 */
void JSMN_ItemInit(jsmn_item_t *item, jsmntok_t *t, int index, int tokens_len)
{
    item->t = t;
    item->index = index;
    item->left_num = tokens_len - index - 1;
}

/**
 * @brief   根据 key 从 object 中获取 Agile Jsmn 对象
 * @param   js json 字符串
 * @param   object object 对象
 * @param   string key 字符串
 * @param   item 存放的 Agile Jsmn 对象指针
 * @return  0:成功; !=0:异常
 */
int JSMN_GetObjectItem(const char *js, jsmn_item_t *object, const char *const string, jsmn_item_t *item)
{
    if ((js == NULL) || (object == NULL) || (string == NULL) || (item == NULL))
        return -1;

    jsmn_item_t ob_item;

    if (object->t->type == JSMN_OBJECT) {
        ob_item.t = object->t;
        ob_item.index = object->index;
        ob_item.left_num = object->left_num;
    } else {
        if ((object->t->type != JSMN_STRING) || (object->t->size != 1) || (object->left_num) <= 0)
            return -1;

        jsmntok_t *t = object->t + 1;
        if ((t->type != JSMN_OBJECT) || (t->parent != object->index))
            return -1;

        ob_item.t = object->t + 1;
        ob_item.index = object->index + 1;
        ob_item.left_num = object->left_num - 1;
    }

    if (ob_item.left_num <= 0)
        return -1;

    for (int i = 1; i <= ob_item.left_num; i++) {
        jsmntok_t *t = ob_item.t + i;
        if ((t->type == JSMN_STRING) && (t->parent == ob_item.index)) {
            if (json_token_streq(js, t, string)) {
                item->t = t;
                item->index = ob_item.index + i;
                item->left_num = ob_item.left_num - i;

                return 0;
            }
        }
    }

    return -1;
}

/**
 * @brief   获取 Agile Jsmn 对象的字符串
 * @note    适用 "key":$value 形式。
 *          - 该 API 会修改 'js' json 字符串，请确保 'js' 参数指向的内存可更改，否则请使用 'JSMN_GetStringBuffered'
 *          - 如果 Agile Jsmn 对象指向的是 key，返回的就是 key 字符串
 *          - 如果 Agile Jsmn 对象指向的是 value，返回的就是 value 值。返回的是字符串，需要自行处理判断
 * @par     $value 示例
 * @code
 *          "key":"123" 形式返回 123
 *          "key":123 形式返回 123
 *          "key":null 形式返回 null
 * @endcode
 * @param   js json 字符串
 * @param   item Agile Jsmn 对象
 * @return  !=NULL:字符串; =NULL:异常
 */
char *JSMN_GetString(char *js, jsmn_item_t *item)
{
    if ((js == NULL) || (item == NULL))
        return NULL;

    if ((item->t->type != JSMN_STRING) && (item->t->type != JSMN_PRIMITIVE))
        return NULL;

    js[item->t->end] = '\0';
    return (js + item->t->start);
}

/**
 * @brief   获取 Agile Jsmn 对象的字符串并存入 buf
 * @note    该 API 功能与 'JSMN_GetString' 相同。如果字符串长度比 'bufsz - 1' 大，则会自动丢弃多余的。
 * @param   js json 字符串
 * @param   item Agile Jsmn 对象
 * @param   buf 存储内存
 * @param   bufsz 存储内存大小
 * @return  >=0:字符串长度; 其他:异常
 */
int JSMN_GetStringBuffered(const char *js, jsmn_item_t *item, char *buf, int bufsz)
{
    memset(buf, 0, bufsz);

    if ((js == NULL) || (item == NULL) || (buf == NULL) || (bufsz <= 0))
        return -1;

    if ((item->t->type != JSMN_STRING) && (item->t->type != JSMN_PRIMITIVE))
        return -1;

    int len = item->t->end - item->t->start;
    if (len >= bufsz)
        len = bufsz - 1;

    memcpy(buf, js + item->t->start, len);

    return len;
}

/**
 * @brief   获取 Agile Jsmn 对象的值字符串
 * @note    适用 "key":$value 形式，Agile Jsmn 对象指向的是 key，获取的是 value 值
 * @see     value 值参见 JSMN_GetString
 * @param   js json 字符串
 * @param   item Agile Jsmn 对象
 * @return  !=NULL:值字符串; =NULL:异常
 */
char *JSMN_GetValueString(char *js, jsmn_item_t *item)
{
    if ((js == NULL) || (item == NULL))
        return NULL;

    if ((item->t->type != JSMN_STRING) || (item->t->size != 1) || (item->left_num <= 0))
        return NULL;

    jsmn_item_t val_item;
    val_item.t = item->t + 1;
    val_item.index = item->index + 1;
    val_item.left_num = item->left_num - 1;

    return JSMN_GetString(js, &val_item);
}

/**
 * @brief   获取 Agile Jsmn 对象的值字符串并存入 buf
 * @note    该 API 功能与 'JSMN_GetValueString' 相同。如果字符串长度比 'bufsz - 1' 大，则会自动丢弃多余的。
 * @param   js json 字符串
 * @param   item Agile Jsmn 对象
 * @param   buf 存储内存
 * @param   bufsz 存储内存大小
 * @return  >=0:字符串长度; 其他:异常
 */
int JSMN_GetValueStringBuffered(const char *js, jsmn_item_t *item, char *buf, int bufsz)
{
    if ((js == NULL) || (item == NULL) || (buf == NULL) || (bufsz <= 0))
        return -1;

    if ((item->t->type != JSMN_STRING) || (item->t->size != 1) || (item->left_num <= 0))
        return -1;

    jsmn_item_t val_item;
    val_item.t = item->t + 1;
    val_item.index = item->index + 1;
    val_item.left_num = item->left_num - 1;

    return JSMN_GetStringBuffered(js, &val_item, buf, bufsz);
}

/**
 * @brief   获取 json 数组成员数目
 * @param   array Agile Jsmn 数组类型对象
 * @return  >=0:数组成员数目; 其他:异常
 */
int JSMN_GetArraySize(jsmn_item_t *array)
{
    if (array == NULL)
        return -1;

    if (array->t->type == JSMN_ARRAY)
        return array->t->size;

    if ((array->t->type != JSMN_STRING) || (array->t->size != 1) || (array->left_num <= 0))
        return -1;

    jsmntok_t *t = array->t + 1;
    if ((t->type != JSMN_ARRAY) || (t->parent != array->index))
        return -1;

    return t->size;
}

/**
 * @brief   根据索引获取 json 数组中的成员对象
 * @param   array Agile Jsmn 数组类型对象
 * @param   index 索引
 * @param   item 存放的 Agile Jsmn 对象指针
 * @return  =0:成功; !=0:异常
 */
int JSMN_GetArrayItem(jsmn_item_t *array, int index, jsmn_item_t *item)
{
    if ((array == NULL) || (index < 0) || (item == NULL))
        return -1;

    if (index >= JSMN_GetArraySize(array))
        return -1;

    jsmn_item_t array_tok;

    if (array->t->type == JSMN_ARRAY) {
        array_tok.t = array->t;
        array_tok.index = array->index;
        array_tok.left_num = array->left_num;
    } else {
        array_tok.t = array->t + 1;
        array_tok.index = array->index + 1;
        array_tok.left_num = array->left_num - 1;
    }

    for (int i = 1; i <= array_tok.left_num; i++) {
        jsmntok_t *t = array_tok.t + i;
        if (t->parent == array_tok.index) {
            if (index > 0) {
                index--;
                continue;
            }
            item->t = t;
            item->index = array_tok.index + i;
            item->left_num = array_tok.left_num - i;

            return 0;
        }
    }

    return -1;
}

/**
 * @}
 */
